File: data-structures-algorithms/python_leetcode/0075-sortColors.py
'''
Difficulty: MEDIUM
No.75    https://leetcode.com/problems/sort-colors/
'''

from typing import List

class Solution:
    def sortColors(self, nums: List[int]) -> None:
        # base case
        if len(nums) == 1: return nums

        # define pointers
        left = 0
        right = 0

        while right < len(nums):
            if nums[right] == 0:
                self.swapElem(nums, left, right)
                left += 1
                right += 1
            else:
                right += 1
        right = left

        while right < len(nums):
            if nums[right] == 1:
                self.swapElem(nums, left, right)
                left += 1
                right += 1
            else:
                right += 1

    def swapElem(self, nums, idx1, idx2):
        temp = nums[idx1]
        nums[idx1] = nums[idx2]
        nums[idx2] = temp

S = Solution()
print(S.sortColors([2, 0, 2, 1, 1, 0])); # [0,0,1,1,2,2]
print(S.sortColors([2, 0, 1])); # [0,1,2]

File: dsa-courses/python-course/js-dsa-revision/0075-sortColors.js
/*
Difficulty: MEDIUM
No.75    https://leetcode.com/problems/sort-colors/

Given an array nums with n objects colored red, white, or blue, 
sort them in-place so that objects of the same color are adjacent, 
with the colors in the order red, white, and blue.

We will use the integers 0, 1, and 2 to represent the color red, white, and blue, respectively.
You must solve this problem without using the library's sort function.

Constraints:

- n == nums.length
- 1 <= n <= 300
- nums[i] is either 0, 1, or 2.
*/

var swap = function (arr, idx1, idx2) {
  var temp = arr[idx1];
  arr[idx1] = arr[idx2];
  arr[idx2] = temp;
};

var sortColors = function (arr) {
  if (arr.length === 1 || arr.length === 0) return arr;

  let left = 0;
  let right = 0;

  while (right < arr.length) {
    if (arr[right] === 0) {
      swap(arr, left, right);
      left++;
      right++;
    } else {
      right++;
    }
  }

  right = left;
  while (right < arr.length) {
    if (arr[right] == 1) {
      swap(arr, left, right);
      left++;
      right++;
    } else {
      right++;
    }
  }
  return arr;
};
// [0,0,2,1,1,2]
console.log(sortColors([2, 0, 2, 1, 1, 0])); // [0,0,1,1,2,2]
console.log(sortColors([2, 0, 1])); // [0,1,2]
